dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/h12/grpc/GrpcCompositeCodec.java

```diff
@@ -16,123 +16,103 @@
 import org.apache.dubbo.remoting.http12.message.HttpMessageCodec;
 import org.apache.dubbo.remoting.http12.message.MediaType;
+import org.apache.dubbo.rpc.model.FrameworkModel;
+import org.apache.dubbo.rpc.model.MethodDescriptor;
+import org.apache.dubbo.rpc.model.PackableMethod;
+import org.apache.dubbo.rpc.model.PackableMethodFactory;
 
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.nio.charset.Charset;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
-import com.google.protobuf.Message;
-
-import static org.apache.dubbo.common.constants.CommonConstants.PROTOBUF_MESSAGE_CLASS_NAME;
+import static org.apache.dubbo.common.constants.CommonConstants.DEFAULT_KEY;
+import static org.apache.dubbo.common.constants.CommonConstants.DUBBO_PACKABLE_METHOD_FACTORY;
 
 public class GrpcCompositeCodec implements HttpMessageCodec {
 
-    private final ProtobufHttpMessageCodec protobufHttpMessageCodec;
+    private static final String PACKABLE_METHOD_CACHE = "PACKABLE_METHOD_CACHE";
 
-    private final WrapperHttpMessageCodec wrapperHttpMessageCodec;
+    private final URL url;
 
-    public GrpcCompositeCodec(
-            ProtobufHttpMessageCodec protobufHttpMessageCodec, WrapperHttpMessageCodec wrapperHttpMessageCodec) {
-        this.protobufHttpMessageCodec = protobufHttpMessageCodec;
-        this.wrapperHttpMessageCodec = wrapperHttpMessageCodec;
-    }
+    private final FrameworkModel frameworkModel;
+
+    private final String mediaType;
 
-    public void setEncodeTypes(Class<?>[] encodeTypes) {
-        this.wrapperHttpMessageCodec.setEncodeTypes(encodeTypes);
+    private PackableMethod packableMethod;
+
+    public GrpcCompositeCodec(URL url, FrameworkModel frameworkModel, String mediaType) {
+        this.url = url;
+        this.frameworkModel = frameworkModel;
+        this.mediaType = mediaType;
     }
 
-    public void setDecodeTypes(Class<?>[] decodeTypes) {
-        this.wrapperHttpMessageCodec.setDecodeTypes(decodeTypes);
+    public void loadPackableMethod(MethodDescriptor methodDescriptor) {
+        if (methodDescriptor instanceof PackableMethod) {
+            packableMethod = (PackableMethod) methodDescriptor;
+            return;
+        }
+        Map<MethodDescriptor, PackableMethod> cacheMap = (Map<MethodDescriptor, PackableMethod>) url.getServiceModel()
+                .getServiceMetadata()
+                .getAttributeMap()
+                .computeIfAbsent(PACKABLE_METHOD_CACHE, k -> new ConcurrentHashMap<>());
+        packableMethod = cacheMap.computeIfAbsent(methodDescriptor, md -> frameworkModel
+                .getExtensionLoader(PackableMethodFactory.class)
+                .getExtension(ConfigurationUtils.getGlobalConfiguration(url.getApplicationModel())
+                        .getString(DUBBO_PACKABLE_METHOD_FACTORY, DEFAULT_KEY))
+                .create(methodDescriptor, url, mediaType));
```

Comment:
**Security Issue**: The implementation of loadPackableMethod lacks input validation for the MethodDescriptor parameter, potentially allowing the insertion of an invalid or malicious MethodDescriptor object. This can lead to security vulnerabilities such as code injection or unauthorized method execution. <br> **Fix**: Validate the MethodDescriptor object to ensure it is not null, and verify that it contains valid data before using it. <br> **Code Suggestion**: 
 ```
 +        if (methodDescriptor == null || !validateMethodDescriptor(methodDescriptor)) {
 +            throw new IllegalArgumentException("Invalid MethodDescriptor provided");
 +        }
 +        Map<MethodDescriptor, PackableMethod> cacheMap = (Map<MethodDescriptor, PackableMethod>) url.getServiceModel()
 +                .getServiceMetadata()
 +                .getAttributeMap()
 +                .computeIfAbsent(PACKABLE_METHOD_CACHE, k -> new ConcurrentHashMap<>());
 +        packableMethod = cacheMap.computeIfAbsent(methodDescriptor, md -> frameworkModel
 +                .getExtensionLoader(PackableMethodFactory.class)
 +                .getExtension(ConfigurationUtils.getGlobalConfiguration(url.getApplicationModel())
 +                        .getString(DUBBO_PACKABLE_METHOD_FACTORY, DEFAULT_KEY))
 +                .create(methodDescriptor, url, mediaType));
 ```

<br> **Performance Issue**: The method 'loadPackableMethod' uses unchecked casting which could lead to ClassCastException at runtime if the object types do not match the expected types. This occurs when casting the result of 'url.getServiceModel().getServiceMetadata().getAttributeMap()' to 'Map<MethodDescriptor, PackableMethod>'. <br> **Fix**: To ensure type safety, consider adding type checks before casting or use Java Generics to enforce type safety at compile time. <br> **Code Suggestion**: 
 ```
 +        Map<MethodDescriptor, PackableMethod> cacheMap = (Map<MethodDescriptor, PackableMethod>) url.getServiceModel()
 +                .getServiceMetadata()
 +                .getAttributeMap()
 +                .computeIfAbsent(PACKABLE_METHOD_CACHE, k -> new ConcurrentHashMap<>());
 -        Map<MethodDescriptor, PackableMethod> cacheMap = url.getServiceModel()
 -                .getServiceMetadata()
 -                .getAttributeMap()
 -                .computeIfAbsent(PACKABLE_METHOD_CACHE, k -> new ConcurrentHashMap<>());
 ```

<br> **Optimization Issue**: Complex method 'loadPackableMethod' with multiple conditional and computational steps affecting readability and maintainability. <br> **Fix**: Refactor 'loadPackableMethod' to separate concerns, possibly by breaking down into smaller methods or by enhancing inline documentation to clarify the process steps. <br> **Code Suggestion**: 
 ```
 +    private void loadPackableMethodRefactored(MethodDescriptor methodDescriptor) {
 +        if (methodDescriptor instanceof PackableMethod) {
 +            this.packableMethod = (PackableMethod) methodDescriptor;
 +            return;
 +        }
 +        initializePackableMethodCache(methodDescriptor);
 +    }
 +
 +    private void initializePackableMethodCache(MethodDescriptor methodDescriptor) {
 +        Map<MethodDescriptor, PackableMethod> cacheMap = initializeCacheMap();
 +        this.packableMethod = createPackableMethod(methodDescriptor, cacheMap);
 +    }
 ```